### PR TITLE
feat(storage): support replicated `LocalHummockStorage`

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Part of https://github.com/risingwavelabs/risingwave/issues/10225.

The temporary storage is needed to replicate upstream shared buffer changes, if the two are not on the same CN.

We add `is_replicated` flag to indicate if a local state store is replicated.

In the case that it is, we will only read its read version when reading via state table.

This is because we have to consider an edge case: when the executor using replicated storage is scheduled to the same CN as the upstream mview executor. If replication happens, we will have duplicate read versions in the shared buffer.

Because local state store read (which happens for state table reads) will ONLY read its own read version, it is fine for state table.

But storage table will read from all read versions (via global state store read), which is an issue since we will have 2 copies (original + replicated).

Hence in this PR we also amend global state store read to de duplicate by ignoring ALL read version replicas.

- [x] Ensure that when we create `ReadVersion`, it should follow the local options.
- [x] Ensure global state store filters out `ReadVersion` which are `replicate`s.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by wenym1 — COMMENTED** [review]
Currently, if we read from state table (with underlying local state store), we only read the `ReadVersion` associated to it, and if we read from storage table (with underlying global state store), it will read all `ReadVersion`s. 

Before merging this PR, we should decide whether the read from the storage table can read the data from the replicated `ReadVersion`s. If allowed, later we should carefully ensure that only the data of vnode from other CNs are replicated to the current CN. If not allowed, in this PR we should distinguish whether a `ReadVersion` is replicated or not, and only return the not replicated `ReadVersion`s to handle global state store read.

**Comment by hzxa21** [comment]
Can we add test cases to ensure that 1) replicated ReadVersion can be seen by the corresponding LocalStateStore (for streaming read) and 2) cannot be seen by HummockStorage (for batch read)?

File: src/storage/src/hummock/state_store.rs
Comment (kwannoel) [inline src/storage/src/hummock/state_store.rs:135]: This filters out replicated read versions, when reading from global state store.

**Comment by kwannoel** [comment]
> Can we add test cases to ensure that 1) replicated ReadVersion can be seen by the corresponding LocalStateStore (for streaming read) and 2) cannot be seen by HummockStorage (for batch read)?

Added them in https://github.com/risingwavelabs/risingwave/pull/10226/commits/95f9337e2c79e6022d365654a4c94d027ef239f2.

**Review by Li0k — APPROVED** [review]
LGTM

**Comment by Li0k** [comment]
@chenzl25 PTAL

**Pull request merged**
